middlewares/accesslog/linear_writer_test.go

```diff
@@ -0,0 +1,60 @@
+package accesslog
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLinearWriter(t *testing.T) {
+	file := logfilePath("")
+
+	w, err := os.Create(file)
+	assert.Nil(t, err, "%v", err)
+	defer w.Close()
+
+	lw := LinearWriter(w)
+	wg := &sync.WaitGroup{}
+
+	n := 10
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func(j int) {
+			fmt.Fprintf(lw, "%d\n", j)
+			wg.Done()
+		}(i)
+	}
+	wg.Wait()
+
+	err = lw.Close()
+	assert.Nil(t, err, "%v", err)
+
+	// read back and verify
+	counts := make([]int, n)
+	r, err := os.Open(file)
+	assert.Nil(t, err, "%v", err)
+	defer r.Close()
```

Comment:
[Deferred function calls are executed in Last In First Out order after the surrounding function returns](https://blog.golang.org/defer-panic-and-recover). So I guess we should swap the order of the two defer statements?